### PR TITLE
Patching ERB rendering of the data-info JSON

### LINKED ISSUE #15916: Follow button on profile is not working json parse error
**Describe the bug**

When clicking on follow button is not working because of a json parse error

**To Reproduce**

1. go to https://dev.to/stack-labs?signin=true
2. notice that you have in error in the console because of a json parse error
3. click on follow
4. its not working

**Expected behavior**

We should be able to follow

**Screenshots**

Screen Shot 2022-01-01 at 4 18 36 PM

**Desktop (please complete the following information):**

- OS, version: macbook m1 pro max
- Browser, version: chrome latest

**Smartphone (please complete the following information):**

- Device:
- OS, version:
- Browser, version:

**Additional context**

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Bug Fix

## Description

Prior to this commit, we were somewhat naively rendering Hash style data
attributes in our ERB templates.  By rendering each hash attribute
separately, we were rendering characters that could break the
javascript (e.g. double hack or backslash `"` or `\`).

By moving to this view_object rendering, we leverage Rails's `to_json`
behavior to ensure properly escaped values.  As part of this exercise, I
generalized the method to allow for other places to benefit from this
behavior.

This generalization also helps ensure that we have a more conformant
rendering (e.g. we should always have an :id, :className, and :name
value in our data-info hash).

_Note: I've updated the user's names for Cypress tests as they are more
likely to catch the particular issue than anything else.  I assume that
I'm going to break some cypress tests and will need some help fixing
them._

## Related Tickets & Documents

Closes #15916, #14704

Supersedes #15983

## QA Instructions, Screenshots, Recordings

How to test locally:

Assuming you have seeded database (e.g. `rails db:seed`), checkout the
"main" branch.  Then in `rails console` find a user that's written articles:

```ruby
user = Article.last.user

user.update(name: "\\: #{user.name}")

user.articles.each(&:save)
```

Now, again on the "main" branch, start your application (e.g.,
`bin/startup`).

Then get a logged in and a logged out browser session going.  Open your
web inspector and open console.  Then go to the local instances homepage
(e.g., http://localhost:3000) and look for JS errors.

On the main branch, you should see an exception around
`JSON.parse(button.data.info)` (assuming that the `user`'s article is
rendered on the homepage).

Then go to the user's page (e.g. https://localhost:3000/:user-slug) and
look for JS parse errors.

On this PR's branch (e.g.,
`jeremyf/take-two-at-resolving-gh-15916`)
you shouldn't see those console errors.

More importantly, the Follow buttons should work.

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] I will share this change internally with the appropriate teams

### CODE REVIEW

File: app/views/stories/tagged_articles/index.html.erb [hunk 1]
```diff
@@ -24,7 +24,7 @@
               <button
                 id="user-follow-butt"
                 class="crayons-btn follow-action-button"
-                data-info='{"id":<%= @tag.id %>,"className":"Tag", "name": "<%= @tag.pretty_name || @tag.name %>"}'>
+                data-info='<% Follows::DataInfo.to_json(object: @tag, class_name: "Tag", name: @tag.pretty_name || @tag.name) %>'>
```
Comment (jgaskins) [inline app/views/stories/tagged_articles/index.html.erb:27]:
```suggestion
                data-info='<%= Follows::DataInfo.to_json(object: @tag, class_name: "Tag", name: @tag.pretty_name || @tag.name) %>'>
```

File: app/services/follows/data_info.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+      kwargs.merge(
+        {
+          id: id,
+          className: class_name,
+          name: name
+        },
+      ).to_json
+    end
+  end
+end
```
Comment (jgaskins) [inline app/services/follows/data_info.rb:13]: How does this module differ from the top-level one below?

**Review by jgaskins — COMMENTED** [review]
Always nice to see copypasta data structures extracted to a generator for the data structures! 💯 

Just a question and a tweak needed.

File: app/services/follows/data_info.rb (hunk 2 above)
Comment (jeremyf) [reply app/services/follows/data_info.rb:13]: Gah, that creature came along for the ride.

**Comment by jeremyf** [comment]
I checked and the failing tests is valid.

```gherkin
Given I am not logged in
When I log in
Then there shouldn't be any JS errors
```

Such a game of whack-a-mole.

**Comment by jeremyf** [comment]
Reading notes:  The error occurs when you are not logged in, then log out and go to `/`.  Once you've encountered the error, if you refresh, you will no longer encounter the error.

**Comment by jeremyf** [comment]
Reading notes: Cypress continues to raise an error, so I'm debugging by way of Cypress.

**Comment by nickytonline** [comment]
Looks like E2E tests are failing with application errors again.

```
Cypress detected that an uncaught error was thrown from a cross origin script.
We cannot provide you the stack trace, line number, or file where this error occurred.
Check your Developer Tools Console for the actual error - it should be printed there.
It's possible to enable debugging these scripts by adding the `crossorigin` attribute and setting a CORS header.
When Cypress detects uncaught errors originating from your application it will automatically fail the current test.
This behavior is configurable, and you can choose to turn this off by listening to the `uncaught:exception` event.
https://on.cypress.io/uncaught-exception-from-application
https://on.cypress.io/cross-origin-script-error
  
  2) Home Feed Navigation
       should navigate to Year view:
     Error: The following error originated from your application code, not from Cypress.
  > Script error.
  ```
  
  Looking into it.

**Comment by jeremyf** [comment]
Reading Note: Some of the erratic failures might be related to the randomness of the feed data with each request.

File: spec/view_objects/data_info_spec.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      let(:parameters) { { object: user } }
+
+      it "parses to valid JSON" do
+        expect(JSON.parse(results))
+          .to eq({ "className" => "User", "id" => 123, "name" => "Duane \"The Rock\" Johnson \\\\.//" })
+      end
+    end
+
+    context "when given a User with additional attributes" do
+      let(:parameters) { { object: user, id: 8_675_309, style: "full" } }
```
Comment (jgaskins) [inline spec/view_objects/data_info_spec.rb:22]: I only just noticed this id. Very nice! :joy:

**Review by jgaskins — APPROVED** [review]
Gorgeous!

**Review by Zhao-Andy — APPROVED** [review]
LGTM!

**Review by citizen428 — APPROVED** [review]

**Comment by jeremyf** [comment]
Supplanted by #16067

**Pull request closed without merging**